vcd/resource_vcd_vapp_nat_rules.go

```diff
@@ -0,0 +1,306 @@
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+
+	vcdClient.lockParentVappWithName(d, vapp.VApp.Name)
+	defer vcdClient.unLockParentVappWithName(d, vapp.VApp.Name)
+
+	err = vapp.RemoveAllNetworkNatRules(d.Get("network_id").(string))
+	if err != nil {
+		log.Printf("[INFO] Error deleting NAT rules: %s", err)
+		return fmt.Errorf("error deleting NAT rules: %s", err)
+	}
+
+	return nil
+}
+
+func resourceVappNetworkNatRulesRead(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+
+	vappNetwork, err := vapp.GetVappNetworkById(d.Get("network_id").(string), false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp network. %s", err)
+	}
+
+	var rules []map[string]interface{}
+	if vappNetwork.Configuration.Features == nil || vappNetwork.Configuration.Features.NatService == nil {
+		log.Print("no Nat rules found.")
+		_ = d.Set("rule", rules)
+	}
+
+	for _, rule := range vappNetwork.Configuration.Features.NatService.NatRule {
+		singleRule := make(map[string]interface{})
+		singleRule["id"] = rule.ID
+		if vappNetwork.Configuration.Features.NatService.NatType == portForwardingNatType {
+			singleRule["external_port"] = rule.VMRule.ExternalPort
+			singleRule["vm_nic_id"] = rule.VMRule.VMNicID
+			singleRule["forward_to_port"] = rule.VMRule.InternalPort
+			singleRule["protocol"] = rule.VMRule.Protocol
+			singleRule["vm_id"] = getVmIdFromVmVappLocalId(vapp, rule.VMRule.VAppScopedVMID)
+		} else if vappNetwork.Configuration.Features.NatService.NatType == ipTranslationNatType {
+			singleRule["vm_nic_id"] = rule.OneToOneVMRule.VMNicID
+			singleRule["external_ip"] = rule.OneToOneVMRule.ExternalIPAddress
+			singleRule["mapping_mode"] = rule.OneToOneVMRule.MappingMode
+			singleRule["vm_id"] = getVmIdFromVmVappLocalId(vapp, rule.OneToOneVMRule.VAppScopedVMID)
+		}
+		rules = append(rules, singleRule)
+	}
+	if vappNetwork.Configuration.Features.NatService.NatType == portForwardingNatType &&
+		vappNetwork.Configuration.Features.NatService.Policy == allowTrafficInPolicy {
+		_ = d.Set("enable_ip_masquerade", true)
+	} else if vappNetwork.Configuration.Features.NatService.NatType == portForwardingNatType &&
+		vappNetwork.Configuration.Features.NatService.Policy == allowTrafficPolicy {
+		_ = d.Set("enable_ip_masquerade", false)
+	}
+	_ = d.Set("nat_type", vappNetwork.Configuration.Features.NatService.NatType)
+	_ = d.Set("rule", rules)
```

Comment:
This error should be checked, as we are setting a complex structure and there is a reasonable possibility that it may fail

Reply:
added cehck